### ISSUE TITLE
Tracking of custom parameters not working

### ISSUE THREAD

**Issue by pardeepolicyx** [mautic/mautic#3558]
| Bug report? | Bug in tracking pixek code with custom parameters.

## Description:
I am trying to add custom parameters in tracking pixel code but it is not working

## If a bug:

| Q   | A
| --- | ---
| Mautic version | 2.6.1
| PHP version | 5.6.28

**Comment by escopecz** [comment]
Please provide the code which you use to test tracking with the custom params.

**Comment by pardeepolicyx** [comment]
<script>
    (function(w,d,t,u,n,a,m){w['MauticTrackingObject']=n;
        w[n]=w[n]||function(){(w[n].q=w[n].q||[]).push(arguments)},a=d.createElement(t),
        m=d.getElementsByTagName(t)[0];a.async=1;a.src=u;m.parentNode.insertBefore(a,m)
    })(window,document,'script','http(s)://yourmautic.com/mtc.js','mt');

     mt('send', 'pageview', {email: 'my@email.com', firstname: 'John'});
</script>

**Comment by pardeepolicyx** [comment]
<script>
   (function(w,d,t,u,n,a,m){w['MauticTrackingObject']=n;
       w[n]=w[n]||function(){(w[n].q=w[n].q||[]).push(arguments)},a=d.createElement(t),
       m=d.getElementsByTagName(t)[0];a.async=1;a.src=u;m.parentNode.insertBefore(a,m)
   })(window,document,'script','http://www.policyx.org/bob/mtc.js','mt');
mt('send', 'pageview', {}, {
    onload: function() {
        alert("working properly");
    },
    onerror: function() {
        alert("error");
    }
});
</script>
 i also use this code tracking is working but onload function is not working and also custom parameters

**Comment by escopecz** [comment]
By using your code, the user was successfully created for me:
![screen shot 2017-03-02 at 09 27 19](https://cloud.githubusercontent.com/assets/1235442/23498998/8d0f75d6-ff2a-11e6-9eaf-edc0285ec219.png)
There must be something else wrong with your setup. Maybe a CORS issue? Any error in the dev tools console?

**Comment by pardeepolicyx** [comment]
There is no error in dev tools console. If this is due to CORS issue then nothing should work but at my end contacts are creating using IP address and othere information saving like page url and page tilte except email and firstname.

**Comment by escopecz** [comment]
Ok. Another idea:

> You can also pass information specific to your contact by setting Mautic contact field(s) to be publicly updatable. 

_from [docs](https://mautic.org/docs/en/contacts/contact_monitoring.html)_

Are the fields you are pushing to publicly updatable?

**Comment by pardeepolicyx** [comment]
how to set publicly updatable?

**Comment by escopecz** [comment]
Go to the contact custom field configuration form and you'll see the option. I'll close this issue since we found the problem.

**Comment by pardeepolicyx** [comment]
Thanks its works now after set publicly updatable

**Comment by pizza-hut** [comment]
Hi,
can I do this?

`<script>
    (function(w,d,t,u,n,a,m){w['MauticTrackingObject']=n;
        w[n]=w[n]||function(){(w[n].q=w[n].q||[]).push(arguments)},a=d.createElement(t),
        m=d.getElementsByTagName(t)[0];a.async=1;a.src=u;m.parentNode.insertBefore(a,m)
    })(window,document,'script','http(s)://yourmautic.com/mtc.js','mt');

     mt('send', 'pageview', {customField: 'value'});
</script>`

It seems like mtc.js can only write to customField when there's an accompanying unique identifier. Am I correct?

**Comment by escopecz** [comment]
You have to set each custom field as "Publicly updatable". Read more in the docs: https://www.mautic.org/docs/en/contacts/contact_monitoring.html

**Comment by pizza-hut** [comment]
Yes, done that but nothing is written. I managed to write to the custom field when I have
`{email:lars@metallica.com,` customField: 'value'}` but not this `{customField: 'value'}`

**Comment by samkristoff** [comment]
Hey, 

I have the same issue, I can only push custom fields when I include the email property. 

This works:
mt('send', 'pageview', { email: "my@email.com", custom_field: 'data');

This does not work:
mt('send', 'pageview', {custom_field: 'data');

Any ideas on how to get the second one to work?

Thanks!

**Comment by stevedrobinson** [comment]
Same here. Is there logic behind this limitation? Should it be outlined more clearly in the documentation? Anyone have a good workaround identified for updating a custom field from the website when you don't have an email address?

**Comment by eschiendorfer** [comment]
Is this still working for you?

`<script>
    (function(w,d,t,u,n,a,m){w['MauticTrackingObject']=n;
        w[n]=w[n]||function(){(w[n].q=w[n].q||[]).push(arguments)},a=d.createElement(t),
        m=d.getElementsByTagName(t)[0];a.async=1;a.src=u;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.genzo.ch/mautic/mtc.js','mt');

    mt('send', 'pageview', {email: 'emanuel@something.com', firstname: 'Mandi'});
</script>`

email and firstname are both public updatable. But the contact remains unkown. It just shows IP and all sites I visit. What's wrong?

**Comment by escopecz** [comment]
@eschiendorfer see https://github.com/mautic/mautic/pull/6802

**Comment by eschiendorfer** [comment]
@escopecz thanks a lot! I can confirm, that it works now :)

**Comment by raghav111jain** [comment]
this script work for get email from URL but i want email with firstname and mobile so how to ?? ASAP
Already enable  **public updatable.** 
```
<script>
	function getUrlParameter(name) {
		name = name.replace(/[\[]/, '\\[').replace(/[\]]/, '\\]');
		var regex = new RegExp('[\\?&]' + name + '=([^&#]*)');
		var results = regex.exec(location.search);
		return results === null ? '' : decodeURIComponent(results[1].replace(/\+/g, ' '));
	};
	
    (function(w,d,t,u,n,a,m){w['MauticTrackingObject']=n;
        w[n]=w[n]||function(){(w[n].q=w[n].q||[]).push(arguments)},a=d.createElement(t),
        m=d.getElementsByTagName(t)[0];a.async=1;a.src=u;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://demo.test.com/test/mtc.js','mt');

	var email  = getUrlParameter('email');

	if(email !== ''){
		mt('send', 'pageview', {email: email});
	} else {
		mt('send', 'pageview');
	}
</script>
```

**Comment by fenosc** [comment]
Hi there guys, I'm having problem with this as well but my case might be a bit more tricky cause I'm using Google Tag Manager.
The screeenshot below shows that the two tags were triggered

- **Mautic DEV** - Triggered on page load
- **Mautic - Pedido Formulario** - Triggered after submitting a form

![image](https://user-images.githubusercontent.com/38872987/94312404-308ae980-ff53-11ea-9f60-939c10c7624e.png)

Also, all variables are properly captured by the datalayer
![image](https://user-images.githubusercontent.com/38872987/94313216-b5c2ce00-ff54-11ea-8549-e9408ac3da2b.png)

Content of the TAGs are:

**Mautic Dev**
`<script>
    (function(w,d,t,u,n,a,m){w['MauticTrackingObject']=n;
        w[n]=w[n]||function(){(w[n].q=w[n].q||[]).push(arguments)},a=d.createElement(t),
        m=d.getElementsByTagName(t)[0];a.async=1;a.src=u;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://mautic.mydomain.com/mtc.js','mt');

    mt('send', 'pageview');
</script>`

**Mautic - Pedido Enviado**
`<script>
      (function(w,d,t,u,n,a,m){w['MauticTrackingObject']=n;
        w[n]=w[n]||function(){(w[n].q=w[n].q||[]).push(arguments)},a=d.createElement(t),
        m=d.getElementsByTagName(t)[0];a.async=1;a.src=u;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://mautic.mydomain.com/mtc.js','mt');

    mt('send',  'pageview', {firstname:'{{Pedido - Primeiro Nome}}',lastname:'{{Pedido - Sobrenome}}',cpf:'{{Pedido - CPF}}', mobile:'{{Pedido - Celular}}', email:'{{Pedido - Email}}',sexo:'{{Pedido - Sexo}}'});
</script>`

Problem is that the contact is not created (or updated) in Mautic.

Here is some information about how everything is set-up

- All of the fields above (firstname, lastname, cpf, mobile, email, sexo) are PUBLICLY UPDATABLE
- I'm not having any CORS errors (I'm tracking visitors IPs in my environment normally)

About the tagmanagers codes

- I've tried to have **Mautic - Pedido Enviado** to have only the "mt()" part, but it didn't work
- I've tried to have **Mautic - Pedido Enviado** have only mt("send", {variables});, but it didn't work (to be clear, I removed the "pageview" parameter.
- I've tried to put tagmanagers variables (example: {{Pedido - Primeiro Nome}}) under quotes and without quotes, both did not work. (example: firstname:'{{Pedido - Primeiro Nome}}'; firstname:{{Pedido - Primeiro Nome}}) )

Thanks a bunch for any help.
